docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+##### Overlayfs
+
+Image layers and writable layers are stored under `/var/lib/docker/overlay/<id>`.
+
+Identical files are hardlinked between images.
+
+The image layers contain all their data under a `root` subdirectory.
+
+Everything under  `/var/lib/docker/overlay/<container-id>` are files required for running the container, including its writable layer.
+
+### Improve disk accounting
+
+Disk accounting is dependent on the storage driver in docker. A common solution that works across all storage drivers isn't available.
+
+I’m listing a few possible solutions for disk accounting below along with their limitations.
+
+We need a plugin model for disk accounting. Some storage drivers in docker will require special plugins.
+
+#### Container Images
+
+As of today, the partition that is holding docker images is flagged by cadvisor, and it uses filesystem stats to identify the overall disk usage of that partition.
+
+Isolated usage of just image layers isn’t available today.
+
+However, continuing to use the entire partition availability for garbage collection purposes in kubelet, should not affect reliability.
+We might garbage collect more often.
+As long as we do not expose features that require persisting old containers, computing image layer usage wouldn’t be necessary.
+
+In case we choose to compute the size of image layers alone, the following are some of the ways to achieve that.
+
+*Note that the strategies mentioned below are applicable in general to other kinds of storage like volumes, etc.*
+
+##### Overlayfs and Aufs
+
+1. `du`
+
+	We can list all the image layer specific directories, excluding container directories, and run `du` on each of those directories.
+
+**Pros**:
+
+* This is the least-intrusive approach.
+
+* It will work off the box without requiring any additional configuration.
+
+**Cons**:
+
+* `du` can consume a lot of cpu and memory. There have been several issues reported against the kubelet in the past that were related to `du`.
+
+* It is time consuming. Cannot be run frequently. Requires special handling to constrain resource usage - setting lower nice value or running in a sub-container.
+
+* Can block container deletion by keeping file descriptors open.
+
+
+2. Linux gid based Disk Quota
+
+	[Disk quota](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Storage_Administration_Guide/ch-disk-quotas.html) feature provided by the linux kernel can be used to track the usage of image layers. Ideally, we need `project` support for disk quota, which let’s us track usage of directory hierarchies using `project ids`. Unfortunately, that feature is only available for zfs filesystems. Since most of our distributions use `ext4` by default, we will have to use either `uid` or `gid` based quota tracking.
+
+	Both `uids` and `gids` are meant for security. Overloading that concept for disk tracking is painful and ugly. But, that is what we have today.
```

Comment:
That (re-using FsGroup) will be a necessity until support for quota
projects gets added to ext4.

On Fri, Nov 6, 2015 at 11:15 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/proposals/disk-accounting.md
> <https://github.com/kubernetes/kubernetes/pull/16889#discussion_r44209211>
> :
>
> > +* It will work off the box without requiring any additional configuration.
> > +
> > +**Cons**:
> > +
> > +* `du` can consume a lot of cpu and memory. There have been several issues reported against the kubelet in the past that were related to `du`.
> > +
> > +* It is time consuming. Cannot be run frequently. Requires special handling to constrain resource usage - setting lower nice value or running in a sub-container.
> > +
> > +* Can block container deletion by keeping file descriptors open.
> > +
> > +
> > +2. Linux gid based Disk Quota
> > +
> > +	[Disk quota](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Storage_Administration_Guide/ch-disk-quotas.html) feature provided by the linux kernel can be used to track the usage of image layers. Ideally, we need `project` support for disk quota, which let’s us track usage of directory hierarchies using `project ids`. Unfortunately, that feature is only available for zfs filesystems. Since most of our distributions use `ext4` by default, we will have to use either `uid` or `gid` based quota tracking.
> > +
> > +	Both `uids` and `gids` are meant for security. Overloading that concept for disk tracking is painful and ugly. But, that is what we have today.
>
> We discussed using FSGroup for local storage quota. Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16889/files#r44209211>.
>